core/src/subject/SDNADecorators.ts

```diff
@@ -98,143 +126,153 @@ export function subjectCollection(opts: CollectionOptions) {
+ }
 
-    descriptor.value = () => {
-        let sdna = ""
-        let subjectName = target.name
-        let obj = new target
-
-        sdna += `subject_class("${subjectName}", c).\n`
-
-        let constructorActions = []
-        if(obj.subjectConstructor && obj.subjectConstructor.length) {
-            constructorActions = constructorActions.concat(obj.subjectConstructor)
-        }
-
-        let instanceConditions = []
-        if(obj.isSubjectInstance && obj.isSubjectInstance.length) {
-            instanceConditions = instanceConditions.concat(obj.isSubjectInstance)
-        }
-
-        let propertiesCode = []
-        let properties = obj.__properties || {}
-        for(let property in properties) {
-            let propertyCode = `property(c, "${property}").\n`
-
-            let { through, initial, required, resolve, resolveLanguage } = properties[property]
-
-            if(resolve) {
-                propertyCode += `property_resolve(c, "${property}").\n`
+ interface SDNAClassOptions {
+    name: string;
+}
 
+export function SDNAClass(opts: SDNAClassOptions) {
+    return function (target: any) {
+        target.prototype.className = opts.name;
+        target.className = opts.name;
+        
+        target.generateSDNA = function() {
+            let sdna = ""
+            let subjectName = opts.name
+            let obj = target.prototype;
+    
+            let uuid = makeRandomPrologAtom(8)
+    
+            sdna += `subject_class("${subjectName}", ${uuid}).\n`
+    
+            let constructorActions = []
+            if(obj.subjectConstructor && obj.subjectConstructor.length) {
+                constructorActions = constructorActions.concat(obj.subjectConstructor)
+            }
+    
+            let instanceConditions = []
+            if(obj.isSubjectInstance && obj.isSubjectInstance.length) {
+                instanceConditions = instanceConditions.concat(obj.isSubjectInstance)
+            }
+    
+            let propertiesCode = []
+            let properties = obj.__properties || {}
+            for(let property in properties) {
+                let propertyCode = `property(${uuid}, "${property}").\n`
+    
+                let { through, initial, required, resolveLanguage, writable } = properties[property]
+    
                 if(resolveLanguage) {
-                    propertyCode += `property_resolve_language(c, "${property}", "${resolveLanguage}").\n`
+                    propertyCode += `property_resolve(${uuid}, "${property}").\n`
+                    propertyCode += `property_resolve_language(${uuid}, "${property}", "${resolveLanguage}").\n`
+                }
+                
+                if(through) {
+                    propertyCode += `property_getter(${uuid}, Base, "${property}", Value) :- triple(Base, "${through}", Value).\n`
+    
+                    if(required) {
+                        instanceConditions.push(`triple(Base, "${through}", _)`)
+                    }    
```

Comment:
Yeah, here we are missing the case the this property is a flag in which case the last triple paramter must also be set to the flag value..